jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/EagerContentHandler.java

```diff
@@ -0,0 +1,618 @@
+                        break;
+                    }
+
+                    // Estimated size is 8 byte framing overhead per chunk plus the chunk size
+                    _estimatedSize += _chunkOverhead + chunk.remaining();
+
+                    boolean oversize = _estimatedSize >= _maxRetainedBytes;
+
+                    if (_reject && oversize && !chunk.isLast())
+                    {
+                        Response.writeError(getRequest(), getResponse(), getCallback(), HttpStatus.PAYLOAD_TOO_LARGE_413);
+                        break;
+                    }
+
+                    if (chunk.isLast() || oversize)
+                    {
+                        if (execute)
+                            getRequest().getContext().execute(this::doHandle);
+                        else
+                            doHandle();
+                        break;
+                    }
+                }
+            }
+
+            @Override
+            public InvocationType getInvocationType()
+            {
+                return InvocationType.NON_BLOCKING;
+            }
+
+            /**
+             * This is run when enough content has been received to dispatch to the next handler.
+             */
+            public void run()
+            {
+                read(true);
+            }
+
+            private void doHandle()
+            {
+                RewindChunksRequest request = new RewindChunksRequest(getRequest(), getCallback(), _chunks);
+                if (!handle(request, getResponse(), request))
+                    request.release();
+            }
+
+            private static class RewindChunksRequest extends Request.Wrapper implements Callback
+            {
+                private final Deque<Content.Chunk> _chunks;
+                private final Callback _callback;
+
+                public RewindChunksRequest(Request wrapped, Callback callback, Deque<Content.Chunk> chunks)
+                {
+                    super(wrapped);
+                    _chunks = chunks;
+                    _callback = callback;
+                }
+
+                @Override
+                public Content.Chunk read()
+                {
+                    if (_chunks.isEmpty())
+                        return super.read();
+                    return _chunks.removeFirst();
+                }
+
+                private void release()
+                {
+                    _chunks.forEach(Content.Chunk::release);
+                    _chunks.clear();
+                }
+
+                @Override
+                public void fail(Throwable failure, boolean last)
+                {
+                    release();
+                    _callback.failed(failure);
+                }
+
+                @Override
+                public void succeeded()
+                {
+                    release();
+                    _callback.succeeded();
+                }
```

Comment:
Override `failed(Throwable)` to release the chunks in case an error response fails.